### PR TITLE
New Adapter: Kargo

### PR DESCRIPTION
**Pull request by ssadman22**
🔗 Link to Documentation: https://github.com/prebid/prebid.github.io/pull/3852

### CODE REVIEW

File: adapters/kargo/params_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	}
+}
+
+var validParams = []string{
+	`{"placementId": ""}`,
+	`{"placementId": "_mx764d"}`,
+}
+
+var invalidParams = []string{
+	`{"placementId": 42}`,
```
Comment (guscarreon) [inline adapters/kargo/params_test.go:42]: Can we include more test scenarios inside `adapters/kargo/params_test.go`'s `invalidParams` array? Maybe something like:
```
   var invalidParams = []string{
       `{"placementId": 42}`,
 +     `{"placementId": }`,
 +     `{}`,
   }
```

File: static/bidder-params/kargo.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "type": "object",
+  
+    "properties": {
+        "placementId": {
+          "type": "string",
+          "description": "Placement ID"
+        }
+      },
+    
+    "required": ["placementId"]
```
Comment (guscarreon) [inline static/bidder-params/kargo.json:14]: Neither `adapters/kargo/kargotest/exemplary/banner.json` nor `adapters/kargo/kargotest/exemplary/native.json` seem to come with a `"placementID"` parameter in any of the `request.imp[i].ext` fields. Are we sure this parameter is required?

File: adapters/kargo/kargo.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type KargoAdapter struct {
+	URI string
```
Comment (guscarreon) [inline adapters/kargo/kargo.go:16]: No need to export this type. We could even rename to `adapter`
```
 - type KargoAdapter struct {
 + type adapter struct {
       URI string
   }
```

File: adapters/kargo/kargo.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
```
Comment (guscarreon) [inline adapters/kargo/kargo.go:66]: Can we add unit test coverage to the error scenarios of the `MakeBids` function by creating JSON tests in the `adapters/kargo/kargotest/suppelemtal/` directory?

Comment (guscarreon) [reply adapters/kargo/kargo.go:66]: image

File: adapters/kargo/kargo.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: getMediaTypeForImp(bid.ImpID, request.Imp),
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
```
Comment (guscarreon) [inline adapters/kargo/kargo.go:76]: Can we add further unit test coverage?
image

File: static/bidder-params/kargo.json (hunk 2 above)
Comment (ssadman22) [reply static/bidder-params/kargo.json:14]: Sorry, this should have been `tagid` instead

File: adapters/kargo/kargo.go (hunk 4 above)
Comment (ssadman22) [reply adapters/kargo/kargo.go:66]: Will do!

File: adapters/kargo/kargo.go (hunk 5 above)
Comment (ssadman22) [reply adapters/kargo/kargo.go:76]: Where can I find prebids sample bid requests? I can add the same `imp.ID` to the exemplary tests and this should be covered

File: adapters/kargo/kargotest/exemplary/banner.json [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+              "ext": {
+              }
+            }
+          ]
+        },
+        "tagid": "73",
+        "secure": 1,
+        "iframebuster": ["ALL"],
+        "ext": {
+        }
```
Comment (guscarreon) [inline adapters/kargo/kargotest/exemplary/banner.json:33]: Exemplary files `adapters/kargo/kargotest/exemplary/banner.json`, `adapters/kargo/kargotest/exemplary/native.json` and `adapters/kargo/kargotest/exemplary/video.json` don't include the `"tagid"` field in their `request.imp[i].ext` fields:
```
 1 {
 2   "mockBidRequest": {
 3     "id": "5f4d1e01",
 4     "at": 1,
 5     "imp": [
 6       {
 7         "id": "8b1bdcca",
 8 *-- 21 lines: "banner": {------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
29         "tagid": "73",
30         "secure": 1,
31         "iframebuster": ["ALL"],
32         "ext": {
33         }
34       }
35     ],
adapters/kargo/kargotest/exemplary/banner.json
```

```
 1 {
 2     "mockBidRequest": {
 3       "id": "5f4d1e01",
 4       "at": 1,
 5       "imp": [
 6           {
 7               "id": "8b1bdcca",
 8 *-- 22 lines: "banner": {------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
30               "tagid": "73",
31               "secure": 1,
32               "iframebuster": [
33                   "ALL"
34               ],
35               "ext": {
36               }
37           }
38       ],
adapters/kargo/kargotest/exemplary/native.json
```

```
 1 {
 2     "mockBidRequest": {
 3       "id": "5f4d1e01",
 4       "at": 1,
 5       "imp": [
 6           {
 7               "id": "8b1bdcca",
 8 +-- 27 lines: "banner": {------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
35               "tagid": "73",
36               "secure": 1,
37               "iframebuster": [
38                   "ALL"
39               ],
40               "ext": {
41               }
42           }
43       ],
adapters/kargo/kargotest/exemplary/video.json
```
Given the fact that `"tagid"` was marked as `required` inside `static/bidder-params/kargo.json`, none of the mock bid requests inside the files above would have made it to your adapter: Prebid Server Core would have discarded them. Can we make them more realistic?

File: adapters/kargo/kargo.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+			case imp.Video != nil:
+				return openrtb_ext.BidTypeVideo
+			case imp.Native != nil:
+				return openrtb_ext.BidTypeNative
+			case imp.Banner != nil:
+				return openrtb_ext.BidTypeBanner
+			}
+		}
+	}
+	return openrtb_ext.BidTypeBanner
```
Comment (guscarreon) [inline adapters/kargo/kargo.go:89]: Strictly speaking, the `return` statement in line 89 makes the last `case` statement unnecessary, but I'm OK either way. I guess an upside of leaving the `case` there to make the function more straightforward.

File: adapters/kargo/kargotest/exemplary/banner.json (hunk 6 above)
Comment (ssadman22) [reply adapters/kargo/kargotest/exemplary/banner.json:33]: This is probably not necessary in `req.imp[i].ext`, but we need either the `tagid`, `placementid` or `adSlotId`

File: adapters/kargo/kargo.go (hunk 7 above)
Comment (ssadman22) [reply adapters/kargo/kargo.go:89]: Thanks for catching that, I can remove it

File: adapters/kargo/kargotest/exemplary/banner.json (hunk 6 above)
Comment (guscarreon) [reply adapters/kargo/kargotest/exemplary/banner.json:25]: Maybe a check can be added inside `MakeRequests` on `req.imp[i].tagid`

Comment (ssadman22) [reply adapters/kargo/kargotest/exemplary/banner.json:25]: Yeah I can add a check there. 
I'm discussing it with my team and we might go with another param such as `AdSlotID` or `AdSlot` for `req.imp[i].ext` instead. I'll confirm everything and re-request a review when its ready.

File: adapters/kargo/kargo.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
```
[the remark this answers is not in the document]
Comment (ssadman22) [reply adapters/kargo/kargo.go:69]: Thanks for the suggestion, will be changing it back! 

Just read up on it and its interesting: 
> The iteration variables may be declared by the "range" clause using a form of [short variable declaration](https://golang.org/ref/spec#Short_variable_declarations) (:=). In this case their types are set to the types of the respective iteration values and their [scope](https://golang.org/ref/spec#Declarations_and_scope) is the block of the "for" statement; they are re-used in each iteration. If the iteration variables are declared outside the "for" statement, after execution their values will be those of the last iteration.

**Comment by guscarreon** [comment]
@ssadman22 there seems to be a problem with the last update. One of your test cases seems to be failing. I didn't take a good look at the details but the first test `"Validate/merge"` could probably need you to incorporate the latest changes in master to your branch.

**Comment by ssadman22** [comment]
> @ssadman22 there seems to be a problem with the last update. One of your test cases seems to be failing. I didn't take a good look at the details but the first test `"Validate/merge"` could probably need you to incorporate the latest changes in master to your branch.

Just saw the email, will post a fix thanks!

**Comment by ssadman22** [comment]
> @ssadman22 there seems to be a problem with the last update. One of your test cases seems to be failing. I didn't take a good look at the details but the first test `"Validate/merge"` could probably need you to incorporate the latest changes in master to your branch.

When I run validation tests, I don't see any failed tests but it exits with Running: go vet. The branch also seems to be up-to-date with master when I tried `git merge`

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**